stream/BidirectionalStream.java

```diff
@@ -0,0 +1,154 @@
+
+public class BidirectionalStream implements AutoCloseable {
+
+    private final ResponseCollector<Transaction.Res> resCollector;
+    private final ResponseCollector<Transaction.ResPart> resPartCollector;
+    private final RequestTransmitter.Dispatcher dispatcher;
+    private final AtomicBoolean isOpen;
+
+    public BidirectionalStream(Channel channel, RequestTransmitter executor) {
+        resPartCollector = new ResponseCollector<>();
+        resCollector = new ResponseCollector<>();
+        isOpen = new AtomicBoolean(false);
+        dispatcher = executor.dispatcher(GraknGrpc.newStub(channel).transaction(responseObserver()));
+        isOpen.set(true);
+    }
+
+    public Single<Transaction.Res> single(Transaction.Req.Builder request, boolean batch) {
+        UUID requestID = UUID.randomUUID();
+        Transaction.Req req = request.setReqId(requestID.toString()).build();
+        ResponseCollector.Queue<Transaction.Res> queue = resCollector.queue(requestID);
+        if (batch) dispatcher.dispatch(req);
+        else dispatcher.dispatchNow(req);
+        return new Single<>(queue);
+    }
+
+    public Stream<Transaction.ResPart> stream(Transaction.Req.Builder request) {
+        UUID requestID = UUID.randomUUID();
+        ResponseCollector.Queue<Transaction.ResPart> collector = resPartCollector.queue(requestID);
+        dispatcher.dispatch(request.setReqId(requestID.toString()).build());
+        ResponseIterator iterator = new ResponseIterator(requestID, collector, dispatcher);
+        return StreamSupport.stream(spliteratorUnknownSize(iterator, ORDERED | IMMUTABLE), false);
+    }
+
+    public boolean isOpen() {
+        return isOpen.get();
+    }
+
+    private void collect(Transaction.Res res) {
+        UUID requestID = UUID.fromString(res.getReqId());
+        ResponseCollector.Queue<Transaction.Res> collector = resCollector.get(requestID);
+        if (collector != null) collector.put(res);
+        else throw new GraknClientException(UNKNOWN_REQUEST_ID.message(requestID));
+    }
+
+    private void collect(Transaction.ResPart resPart) {
+        UUID requestID = UUID.fromString(resPart.getReqId());
+        ResponseCollector.Queue<Transaction.ResPart> collector = resPartCollector.get(requestID);
+        if (collector != null) collector.put(resPart);
+        else throw new GraknClientException(UNKNOWN_REQUEST_ID.message(requestID));
+    }
+
+    @Override
+    public void close() {
+        close(null);
+    }
+
+    private void close(@Nullable StatusRuntimeException error) {
+        if (isOpen.compareAndSet(true, false)) {
+            resCollector.close(error);
+            resPartCollector.close(error);
+            try {
+                dispatcher.close();
+            } catch (StatusRuntimeException e) {
+                throw GraknClientException.of(e);
+            }
+        }
+    }
+
+    private StreamObserver<Transaction.Server> responseObserver() {
```

Comment:
I find these clearer to follow when we don't use aliases - ie:
```suggestion
    private StreamObserver<TransactionProto.Transaction.Server> responseObserver() {
```